clients/include/argument_model.hpp

```diff
@@ -0,0 +1,152 @@
+/* ************************************************************************
+ * Copyright 2020 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+
+#ifndef _ARGUMENT_MODEL_HPP_
+#define _ARGUMENT_MODEL_HPP_
+
+#include "rocblas_arguments.hpp"
+
+// ArgumentModel template has a variadic list of argument enums
+template <rocblas_argument... Args>
+class ArgumentModel
+{
+    // Whether model has a particular parameter (compile-time)
+    static constexpr bool has(rocblas_argument param)
+    {
+        // TODO: Replace with C++17 fold expression
+        for(auto x : {Args...})
+            if(x == param)
+                return true;
+        return false;
+    }
+
+    template <rocblas_argument>
+    struct apply
```

Comment:
This class should not be nested in here as it can ruin any memorizing by the compiler because each `apply` is dependent on the outer `Args...` even though they are not used by the `apply` class.

Reply:
Some functions rely on the template arguments, which is why it is a nested class.

Reply:
Originally I tried separating it into `.hpp/.cpp` files, but then I found that the instantiations required would be too long, and would be complicated by the fact that lambdas (captures) have unique types which can't be expressed in another file, only in the same file with `decltype()`. So I had to put them in the header, and since they sometimes depend on template parameters, I had to make it a nested class.

Reply:
For some reason I thought it depended on template parameters, but I guess not. I moved it to a totally separate non-templated class.